### PR TITLE
ARROW-6610: [C++] Add cmake option to disable filesystem layer

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
@wesm @kou

File: cpp/src/arrow/CMakeLists.txt [hunk 1]
```diff
@@ -228,48 +219,38 @@ if(ARROW_COMPUTE)
...
-  set(ARROW_SRCS util/compression_snappy.cc ${ARROW_SRCS})
-endif()
+if(ARROW_FILESYSTEM)
+  add_subdirectory(filesystem)
 
-if(ARROW_WITH_ZLIB)
-  add_definitions(-DARROW_WITH_ZLIB)
-  set(ARROW_SRCS util/compression_zlib.cc ${ARROW_SRCS})
-endif()
+  set(ARROW_SRCS
```
Comment (kou) [inline cpp/src/arrow/CMakeLists.txt]: Could you use `list(APPEND ARROW_SRCS ...)`?
https://cmake.org/cmake/help/latest/command/list.html#modification

**Review by kou — APPROVED** [review]
I'm not sure this patch doesn't cause any regression because this patch has many unrelated changes.
But this patch will not cause any regression because our tests don't detect any regression.

I think that we can merge this once we change `set(XXX ${XXX} ...)` to `list(APPEND XXX ...)`.

File: cpp/src/arrow/CMakeLists.txt (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/CMakeLists.txt]: Done.

**Comment by pitrou** [comment]
(the "unrelated" changes are just reorganizing the CMakeLists.txt to make ordering more logical)

**Comment by kszucs** [comment]
@pitrou this means that we need to make the python and R filesystem bindings optional as well.

**Pull request closed without merging**